In [ ]:
# imports
from pathlib import Path
import subprocess
import sys
import pandas as pd

# Explore Upper-Limb Recordings

This notebook inspects the upper-limb recordings without changing the raw data.

We begin with the inventory already created at `data/inventories/upper_limb_inventory.csv`.

In [15]:
project_root = Path.cwd().parent
recordings_root = project_root / "data" / "upper_limb_recorded_data" / "upper_limb_asessment" / "recordings"
inventory_path = project_root / "data" / "inventories" / "upper_limb_inventory.csv"
inventory_script = project_root / "imkin" / "inventory_upper_limb.py"

if not inventory_path.exists():
    subprocess.run(
        [
            sys.executable,
            str(inventory_script),
            "--root",
            str(recordings_root),
            "--out",
            str(inventory_path),
        ],
        check=True,
    )
else:
    print(f"Using existing inventory: {inventory_path}")

inventory = pd.read_csv(inventory_path)

print(f"Inventory path: {inventory_path}")
print(f"Recordings in inventory: {len(inventory)}")
inventory.head()

Found 112 upper-limb recordings: 106 complete, 6 incomplete.
Wrote /Users/louistschanz/Documents/EPFL/Champalimaud/immersive-kinematics/data/inventories/upper_limb_inventory.csv
Inventory path: /Users/louistschanz/Documents/EPFL/Champalimaud/immersive-kinematics/data/inventories/upper_limb_inventory.csv
Recordings in inventory: 112


,recording_id,status,events_path,frames_path,meta_path,events_bytes,frames_bytes,meta_bytes,events_lines,frames_data_rows
0,upper_limb_20260629_090253,complete,upper_limb_20260629_090253.events.jsonl,upper_limb_20260629_090253.frames.csv,upper_limb_20260629_090253.meta.json,36645,486646,5085,57,453
1,upper_limb_20260629_090948,complete,upper_limb_20260629_090948.events.jsonl,upper_limb_20260629_090948.frames.csv,upper_limb_20260629_090948.meta.json,21601,696710,5084,32,829
2,upper_limb_20260629_091039,complete,upper_limb_20260629_091039.events.jsonl,upper_limb_20260629_091039.frames.csv,upper_limb_20260629_091039.meta.json,115927,1360877,5085,159,1265
3,upper_limb_20260629_095022,complete,upper_limb_20260629_095022.events.jsonl,upper_limb_20260629_095022.frames.csv,upper_limb_20260629_095022.meta.json,291979,17145542,5085,472,14424
4,upper_limb_20260629_100413,complete,upper_limb_20260629_100413.events.jsonl,upper_limb_20260629_100413.frames.csv,upper_limb_20260629_100413.meta.json,118456,4124721,5085,168,3416


In [16]:
# Check event/frame pairing and report the two non-empty recordings without frames
recordings_root = project_root / "data" / "upper_limb_recorded_data" / "upper_limb_asessment" / "recordings"

# Keep this check aligned with inventory_upper_limb.py: calibration logs are excluded.
event_ids = {
    path.name.removesuffix(".events.jsonl")
    for path in recordings_root.glob("upper_limb_*.events.jsonl")
}
frame_ids = {
    path.name.removesuffix(".frames.csv")
    for path in recordings_root.glob("upper_limb_*.frames.csv")
}

events_without_frames = sorted(event_ids - frame_ids)
frames_without_events = sorted(frame_ids - event_ids)

strange = inventory[
    (inventory["events_lines"] > 0) & (inventory["frames_data_rows"] == 0)
][["recording_id", "events_lines", "frames_data_rows", "status"]]

print(f"Inventory recordings: {len(inventory)}")
print(f"Event recordings checked: {len(event_ids)}")
print(f"Frame recordings checked: {len(frame_ids)}")
print(f"Event files without a frame file: {events_without_frames}")
print(f"Frame files without an event file: {frames_without_events}")
print(f"Paired event/frame recordings: {len(event_ids & frame_ids)}")
print("\nTwo strange recordings: events exist, but no frame rows were recorded")
print(strange.to_string(index=False))

assert len(event_ids) == len(inventory)
assert len(frame_ids) == len(inventory)
assert not events_without_frames
assert not frames_without_events
assert len(strange) == 2

Inventory recordings: 112
Event recordings checked: 112
Frame recordings checked: 112
Event files without a frame file: []
Frame files without an event file: []
Paired event/frame recordings: 112

Two strange recordings: events exist, but no frame rows were recorded
              recording_id  events_lines  frames_data_rows     status
upper_limb_20260728_134516             2                 0 incomplete
upper_limb_20260730_132157             3                 0 incomplete


In [17]:
calibration_inventory_path = project_root / "data" / "inventories" / "calibration_inventory.csv"
calibration_inventory_script = project_root / "imkin" / "inventory_calibration.py"

if not calibration_inventory_path.exists():
    subprocess.run(
        [
            sys.executable,
            str(calibration_inventory_script),
            "--root",
            str(recordings_root),
            "--out",
            str(calibration_inventory_path),
        ],
        check=True,
    )
else:
    print(f"Using existing calibration inventory: {calibration_inventory_path}")

calibration_inventory = pd.read_csv(calibration_inventory_path)

print(f"Calibration inventory path: {calibration_inventory_path}")
print(f"Calibration recordings: {len(calibration_inventory)}")
calibration_inventory.head()

Found 30 calibration recordings: 29 complete, 1 incomplete.
Wrote /Users/louistschanz/Documents/EPFL/Champalimaud/immersive-kinematics/data/inventories/calibration_inventory.csv
Calibration inventory path: /Users/louistschanz/Documents/EPFL/Champalimaud/immersive-kinematics/data/inventories/calibration_inventory.csv
Calibration recordings: 30


,recording_id,calibration_type,status,events_path,frames_path,meta_path,events_bytes,frames_bytes,meta_bytes,events_lines,frames_data_rows
0,r1_calibration_20260616_173423,calibration,complete,r1_calibration_20260616_173423.events.jsonl,r1_calibration_20260616_173423.frames.csv,r1_calibration_20260616_173423.meta.json,16227,756739,1702,11,835
1,r1_calibration_20260616_173530,calibration,complete,r1_calibration_20260616_173530.events.jsonl,r1_calibration_20260616_173530.frames.csv,r1_calibration_20260616_173530.meta.json,13531,129985,1701,11,151
2,r1_calibration_20260616_173543,calibration,complete,r1_calibration_20260616_173543.events.jsonl,r1_calibration_20260616_173543.frames.csv,r1_calibration_20260616_173543.meta.json,9584,890023,1701,7,859
3,r1_calibration_20260616_173731,calibration,complete,r1_calibration_20260616_173731.events.jsonl,r1_calibration_20260616_173731.frames.csv,r1_calibration_20260616_173731.meta.json,9615,635629,1701,7,568
4,r1_calibration_20260616_173756,calibration,complete,r1_calibration_20260616_173756.events.jsonl,r1_calibration_20260616_173756.frames.csv,r1_calibration_20260616_173756.meta.json,17842,894767,1702,11,770


## What is inside one recording?

Each recording `upper_limb_YYYYMMDD_HHMMSS` is made of three files:

- `*.frames.csv`: one row per pose frame (tracked body keypoints over time).
- `*.events.jsonl`: one JSON object per line, one line per event (task/trial starts and ends, targets reached, ...). Each event `type` has its own keys.
- `*.meta.json`: one nested dict with the session configuration (services, pointing model, task thresholds).

Below we open one complete recording and list the column names / keys of each file.

In [ ]:
# Pick one complete recording as an example
import json

example_id = "upper_limb_20260629_095022"
example_frames_path = recordings_root / f"{example_id}.frames.csv"
example_events_path = recordings_root / f"{example_id}.events.jsonl"
example_meta_path = recordings_root / f"{example_id}.meta.json"

print(inventory.loc[inventory["recording_id"] == example_id].T.to_string(header=False))

### `frames.csv`

The first 12 columns describe the session and trial state for each frame. The remaining 60 are 15 body keypoints × (`_x`, `_y`, `_z`, `_conf`).

In [ ]:
example_frames = pd.read_csv(example_frames_path)

keypoint_suffixes = ("_x", "_y", "_z", "_conf")
info_columns = [c for c in example_frames.columns if not c.endswith(keypoint_suffixes)]
keypoints = list(dict.fromkeys(
    c.rsplit("_", 1)[0] for c in example_frames.columns if c.endswith(keypoint_suffixes)
))
keypoint_groups = {
    "Head": ["nose", "left_eye", "right_eye", "left_ear", "right_ear"],
    "Trunk": ["neck", "body_center", "left_hip", "right_hip"],
    "Left arm": ["left_shoulder", "left_elbow", "left_wrist"],
    "Right arm": ["right_shoulder", "right_elbow", "right_wrist"],
}
assert sorted(sum(keypoint_groups.values(), [])) == sorted(keypoints)

print(f"{example_id}.frames.csv")
print(f"{len(example_frames):,} rows (one per pose frame) x {len(example_frames.columns)} columns")
print(f"  = {len(info_columns)} session/trial columns")
print(f"  + {len(keypoints)} keypoints x {len(keypoint_suffixes)} values {keypoint_suffixes}")

print(f"\n1) Session / trial columns ({len(info_columns)})")
for c in info_columns:
    print(f"   - {c}")

print(f"\n2) Keypoints ({len(keypoints)}), each stored as <keypoint>_x, _y, _z, _conf")
for group, names in keypoint_groups.items():
    print(f"   {group:<10} {', '.join(names)}")

# Check that every upper-limb recording shares the same frames header
headers = {
    path.open().readline().strip()
    for path in recordings_root.glob("upper_limb_*.frames.csv")
}
print(f"\nDistinct frames headers across all upper-limb recordings: {len(headers)}")

# Show all 72 columns (pandas hides the middle ones with '...' by default)
with pd.option_context("display.max_columns", None):
    display(example_frames.head())

### `events.jsonl`

There is no single header here: each event `type` carries its own keys. Almost every event has `type` and `wall_time`, and trial-related events also have `task_id, task_name, side, trial_index, trial_count`.

The `session_start` event holds the same content as `meta.json`, so its keys are shown in the meta section below.

In [ ]:
def read_events(path):
    with path.open() as f:
        return [json.loads(line) for line in f if line.strip()]


def summarize_event_types(events):
    """Count events per type and collect the keys seen for each type (in first-seen order)."""
    counts = {}
    keys = {}
    for event in events:
        event_type = event.get("type")
        counts[event_type] = counts.get(event_type, 0) + 1
        keys.setdefault(event_type, {}).update(dict.fromkeys(event))
    return (
        pd.DataFrame({
            "type": list(counts),
            "count": list(counts.values()),
            "keys": [", ".join(keys[t]) if t != "session_start" else "<same as meta.json>" for t in counts],
        })
        .sort_values("count", ascending=False)
        .reset_index(drop=True)
    )


example_events = read_events(example_events_path)
print(f"Events in {example_id}: {len(example_events)}")

with pd.option_context("display.max_colwidth", None):
    display(summarize_event_types(example_events))

#### How to read `events.jsonl`: a timeline

Every line has a `type`. There are two kinds of lines:

- **One big line**, the first one: `session_start`. It holds the session settings (a copy of `meta.json`).
- **Many small lines**: one per thing that happened, in time order, each with its own keys.

The small lines are nested: **session → task → trial → steps of the trial**. A typical trial looks like:

`trial_start` → `trial_start_pose_locked` (patient in start pose) → `target_reached` → `start_returned` (arm back to start) → `trial_end`

The cell below prints the first events of the example recording, indented by level.

In [ ]:
def print_timeline(events, max_events=30):
    """Print events in time order, indented by level (session > task > trial > step)."""
    t0 = events[0]["wall_time"]
    for event in events[:max_events]:
        event_type = event["type"]
        if event_type.startswith("session") or "task_id" not in event:
            indent, label = "", ""
        elif event_type.startswith(("task_", "countdown_")):
            indent, label = "  ", f"task {event['task_id']}"
        elif event_type in ("trial_start", "trial_end"):
            indent, label = "    ", f"task {event['task_id']}, trial {event['trial_index']}"
        else:
            indent, label = "      ", ""

        extra = f"  ({event['duration_s']:.1f} s, {event['reason']})" if event_type == "trial_end" else ""
        size = f"[{len(event)} keys]"
        print(f"{event['wall_time'] - t0:7.1f} s  {size:<10} {indent}{event_type:<26} {label}{extra}")
    if len(events) > max_events:
        print(f"... {len(events) - max_events} more events")


print_timeline(example_events)

One recording does not contain every event type, so we also list the event types across all upper-limb recordings.

In [ ]:
all_events = [
    event
    for path in sorted(recordings_root.glob("upper_limb_*.events.jsonl"))
    for event in read_events(path)
]
print(f"Events across all upper-limb recordings: {len(all_events)}")

with pd.option_context("display.max_colwidth", None, "display.max_rows", None):
    display(summarize_event_types(all_events))

### `meta.json`

A nested dict with the session configuration. We flatten it into `parent.child` keys so every field is visible along with its example value.

In [ ]:
def flatten(obj, prefix=""):
    """Flatten nested dicts into {'a.b.c': value}; lists are kept as values."""
    if not isinstance(obj, dict):
        return {prefix: obj}
    flat = {}
    for key, value in obj.items():
        flat.update(flatten(value, f"{prefix}.{key}" if prefix else key))
    return flat


with example_meta_path.open() as f:
    meta = json.load(f)

meta_flat = pd.Series(flatten(meta), name="value").sort_index()
print(f"Top-level keys: {len(meta)}   Flattened fields: {len(meta_flat)}")

with pd.option_context("display.max_rows", None, "display.max_colwidth", 80):
    display(meta_flat.to_frame())

## Summary of the whole dataset

**Where the data is.** Everything lives in one folder, `data/upper_limb_recorded_data/upper_limb_asessment/recordings/`. The filename prefix tells you what a file is:

| filename starts with | what it is |
|---|---|
| `upper_limb_` | patient sessions, **the data we analyse** |
| `r1_calibration`, `r1_table_` | calibration of the setup (pointing, table), not patient data |
| `adaptive_pointing_` | a different experiment (single `.jsonl` files) |

**One session = 3 files** with the same name `upper_limb_<date>_<time>`:
- `.frames.csv`: body pose ~17 times per second (one row = one snapshot of 15 body points in 3D)
- `.events.jsonl`: what happened and when (task start/end, trial start/end, target reached, ...)
- `.meta.json`: session settings

**Linking them.** Frames and events both have `wall_time`. Frames also carry `active_task_id` and `trial_index`, so you can cut the movement into tasks and trials.

The tables below summarise every upper-limb session and every task.

In [ ]:
# Files in the recordings folder, grouped by kind
all_files = [p.name for p in recordings_root.iterdir() if p.is_file() and not p.name.startswith(".")]


def file_kind(name):
    if name.startswith("upper_limb_"):
        return "upper_limb (patient sessions)"
    if name.startswith("r1_"):
        return "r1 (calibration)"
    if name.startswith("adaptive_pointing_"):
        return "adaptive_pointing (other experiment)"
    return "other"


def file_ext(name):
    for ext in (".frames.csv", ".events.jsonl", ".meta.json", ".jsonl"):
        if name.endswith(ext):
            return ext
    return Path(name).suffix


folder_overview = (
    pd.DataFrame({"kind": [file_kind(n) for n in all_files], "file": [file_ext(n) for n in all_files]})
    .value_counts()
    .unstack(fill_value=0)
)
folder_overview["total files"] = folder_overview.sum(axis=1)
folder_overview

### All upper-limb sessions

One row per session: who, when, how long, which tasks, and how many trials.

In [ ]:
# One row per upper-limb session
session_rows = []
task_names = {}

for recording_id in inventory["recording_id"]:
    events = read_events(recordings_root / f"{recording_id}.events.jsonl")
    frame_times = pd.read_csv(recordings_root / f"{recording_id}.frames.csv", usecols=["wall_time"])["wall_time"]

    patient_id = ""
    dominant_arm = ""
    tasks_done = []
    trials_done = 0
    for event in events:
        if event.get("patient_id"):
            patient_id = event["patient_id"]
        if event.get("dominant_arm"):
            dominant_arm = event["dominant_arm"]
        if event["type"] == "task_start":
            task_names[event["task_id"]] = event["task_name"]
            if event["task_id"] not in tasks_done:
                tasks_done.append(event["task_id"])
        if event["type"] == "trial_end":
            trials_done += 1

    duration_s = frame_times.iloc[-1] - frame_times.iloc[0] if len(frame_times) > 1 else 0.0
    date, time = recording_id.removeprefix("upper_limb_").split("_")
    session_rows.append({
        "recording_id": recording_id,
        "date": pd.to_datetime(date).date(),
        "start": f"{time[:2]}:{time[2:4]}",
        "patient_id": patient_id,
        "dominant_arm": dominant_arm,
        "duration_min": round(duration_s / 60, 1),
        "frames": len(frame_times),
        "fps": round(len(frame_times) / duration_s, 1) if duration_s else 0.0,
        "events": len(events),
        "n_tasks": len(tasks_done),
        "tasks": ", ".join(str(t) for t in sorted(tasks_done)),
        "trials": trials_done,
        "has_meta": (recordings_root / f"{recording_id}.meta.json").exists(),
    })

sessions = pd.DataFrame(session_rows)

print(f"Upper-limb sessions: {len(sessions)}")
print(f"Recording days:      {sessions['date'].nunique()} ({sessions['date'].min()} -> {sessions['date'].max()})")
print(f"Named patient IDs:   {sessions['patient_id'].replace('', pd.NA).nunique()}  (empty = not entered)")
print(f"Total recorded time: {sessions['duration_min'].sum() / 60:.1f} h")
print(f"Total frames:        {sessions['frames'].sum():,}")
print(f"Total trials:        {sessions['trials'].sum():,}")
print(f"Sessions without any trial: {(sessions['trials'] == 0).sum()}")
print(f"Sessions without meta.json: {(~sessions['has_meta']).sum()}")

with pd.option_context("display.max_rows", None, "display.max_columns", None):
    display(
        sessions.style
        .hide(axis="index")
        .format({"duration_min": "{:.1f}", "fps": "{:.1f}", "frames": "{:,}"})
        .bar(subset=["duration_min", "trials"], color="#9ecae1")
    )

### The tasks

How often each task was run, over how many sessions, and how many trials it produced in total.

In [ ]:
task_rows = []
for recording_id in inventory["recording_id"]:
    for event in read_events(recordings_root / f"{recording_id}.events.jsonl"):
        if event["type"] in ("task_start", "trial_end"):
            task_rows.append({"recording_id": recording_id, "type": event["type"], "task_id": event["task_id"]})
task_events = pd.DataFrame(task_rows)

tasks = pd.DataFrame({
    "task_name": pd.Series(task_names),
    "sessions": task_events[task_events["type"] == "task_start"].groupby("task_id")["recording_id"].nunique(),
    "times_run": task_events[task_events["type"] == "task_start"].groupby("task_id").size(),
    "trials": task_events[task_events["type"] == "trial_end"].groupby("task_id").size(),
}).fillna(0).astype({"sessions": int, "times_run": int, "trials": int}).sort_index()
tasks.index.name = "task_id"

tasks.style.bar(subset=["sessions", "trials"], color="#9ecae1")

## Data structure cheat sheet

### 1. Files

Folder: `data/upper_limb_recorded_data/upper_limb_asessment/recordings/`

| prefix | what | count |
|---|---|---|
| `upper_limb_<date>_<time>` | **patient sessions (the data)** | 112 sessions |
| `r1_calibration…`, `r1_table_…` | setup calibration, not patient data | 31 recordings |
| `adaptive_pointing_…` | another experiment | 31 `.jsonl` + 21 `.png` |

One session is made of 3 files with the same name:

| file | format | one row / line = | what it tells you |
|---|---|---|---|
| `.frames.csv` | CSV table, 72 columns | one snapshot of the body (~17 per second) | **the movement**: 3D positions of 15 body points |
| `.events.jsonl` | one JSON object per line | one thing that happened | **what happened and when**: tasks, trials, targets |
| `.meta.json` | one nested JSON dict | the whole session | **settings**: a convenience copy of the `session_start` event, missing for 4 sessions. **`session_start` is the reference.** |

### 2. `frames.csv` (72 columns = 12 + 15 × 4)

| group | columns |
|---|---|
| time | `pose_t` (when the cameras saw the pose, **use for movement**), `wall_time` (when the software saved it, **use to match events**), `pose_age_s` (= `wall_time − pose_t`), `pose_fresh`, `frame_index` |
| who | `session_id`, `person_id` (**tracker ID, not the patient**), `people_count`, `selected_side` |
| what is going on | `active_task_id` (empty = no task running), `trial_active` (1 = inside a trial), `trial_index` |
| head (4 × 5) | `nose`, `left_eye`, `right_eye`, `left_ear`, `right_ear` |
| trunk (4 × 4) | `neck`, `body_center`, `left_hip`, `right_hip` |
| left arm (4 × 3) | `left_shoulder`, `left_elbow`, `left_wrist` |
| right arm (4 × 3) | `right_shoulder`, `right_elbow`, `right_wrist` |

Each keypoint has 4 columns: `<keypoint>_x`, `_y`, `_z` (metres) and `_conf` (tracking confidence).

**Three traps:**

- **`pose_t` vs `wall_time`.** The cameras capture at a steady rhythm (`pose_t`, ~0.056 s between frames). The task software receives the poses irregularly (`wall_time` gaps range from 0.0004 s to 1 s). Speeds computed with `wall_time` can be wrong by a factor of 100, so always use `pose_t` for kinematics.
- **`person_id` is not the patient.** It is the number the tracker gives to the body it sees, and it changes when tracking is lost or someone else walks in (one session has 13 different values). The patient is the `patient_id` typed by the operator during the session (`patient_id_set` event). The `patient_id` field in `session_start` is always empty, and only sessions from 29 July onwards have an ID.
- **Blank cells mean "not recorded", not zero.** Keep them as NaN.

### 3. `events.jsonl` (37 event types, nested session → task → trial → step)

Every event has `type` and `wall_time`.
**trial ctx** = `task_id, task_name, side, trial_index, trial_count`. These are present on every trial-level event.
*Optional* keys only appear for some tasks or in newer recordings, so read them with `event.get(...)`.

| level | type | count | keys (besides `type`, `wall_time`) |
|---|---|---|---|
| **session** | `session_start` | 108 | ~45–60 settings (reference copy; `meta.json` duplicates it; grew over time) |
| | `session_end` | 105 | `session_id, frames` |
| | `patient_id_set` / `dominant_arm_set` | 32 / 33 | `patient_id` / `dominant_arm`, `source`: **the only place where the patient ID and arm are recorded** |
| | `arm_select` | 145 | `side` |
| **task** | `task_start` | 256 | `task_id, task_name, side, trial_count` |
| | `countdown_start` / `countdown_end` | 256 / 248 | `task_id, task_name, side` (+ `duration_s`) |
| | `task_end` | 256 | `task_id, task_name, side, trial_count, completed_trials, duration_s, reason` |
| **trial** | `trial_start` | 3013 | trial ctx + `trial_spec` |
| | `trial_end` | 3013 | trial ctx + `duration_s, reason` |
| **trial step** | `trial_start_pose_locked` | 2975 | trial ctx + `shoulder_wrist_vector, elbow_angle_deg, frozen_trial_spec, …` *(optional: task2_/task6_/task8_ thresholds)* |
| | `target_reached` | 2814 | trial ctx + `target, excursion_m` *(optional, depends on task: `aim_uv, distance_px` pointing · `sweep_rad` circle · `mouth_*` drinking · `nose_*` index-to-nose)* |
| | `path_cue_reached` | 1274 | trial ctx + `stage_index, stage_count, target, aim_uv, distance_px, excursion_m` (task 3) |
| | `start_returned` | 1790 | trial ctx + `return_distance_m, arm_down_ok, elbow_angle_deg, shoulder_displacement_m, …` |
| | `circle_completed` | 201 | trial ctx + `circle, radius_u, radial_error_px, sweep_rad, …` (task 5) |
| | `task6_stage_change` | 537 | trial ctx + `stage` *(optional: distances, dwell)* (task 7, despite the name) |
| | `cup_reached` / `cup_returned` | 110 / 77 | trial ctx + `wrist_position, cup_position, cup_distance_m, …` (task 7) |
| | `mouth_reached` | 246 | trial ctx + `wrist_position, mouth_position, mouth_distance_m, …` (task 7) |
| | `nose_reached` / `nose_left_a` | 243 / 267 | trial ctx + `wrist_position, nose_position, nose_distance_m, …` (task 8) |
| | `initial_a_wait_start` / `initial_a_reached` | 88 / 77 | trial ctx + `anchor_x_m, anchor_z_m, …` (table start point, tasks 6–8) |
| | `target_timeout`, `returned_to_a_without_target` | 3 / 3 | trial ctx + `target, elapsed_s, …` (failed trial) |
| **operator** | `manual_trial_advance` (+ `_ignored`) | 1149 / 970 | trial ctx + `trial_active, trial_spec` *(optional: `key, trigger, r1_press`)* |
| | `r1_a_press`, `manual_initial_a_advance`, `session_finish_requested` | 37 / 4 / 44 | trial ctx + press / trigger info |
| **system** | `projector_status`, `r1_status`, `frame_recorder_stop` | 112 / 2 / 44 | `status` (+ error info) |
| | `runtime_blocker` / `_cleared` | 41 / 21 | `code, severity, summary, detail` |
| | `floor_calibration_triangle_toggle` | 32 | `enabled, source` |

Counts are summed over all 112 sessions.

### 4. `meta.json`

A nested dict with the session settings, grouped as:

| group | example keys |
|---|---|
| session | `session_id, patient_id, dominant_arm, dominant_eye, recording_format` |
| file paths | `events_path, frames_path, meta_path` |
| pointing model | `pointing_ray_mode, pointing_correction{global, per_side{left,right}}` |
| services | `pose_*`, `projector_*`, `frame_buffer_*` |
| task thresholds | `mouth_reach_radius_m, nose_*, task4_*, task6_*, task8_*, table_*` |

### 5. How to cut the movement into trials

`trial_start.wall_time` → `trial_end.wall_time` gives the time window of a trial. Select the frames with `wall_time` inside that window (events and frames share `wall_time`), or use `active_task_id` + `trial_index` + `trial_active == 1` in the frames directly. Once a trial's frames are selected, compute durations and speeds with **`pose_t`**.

Source: `DATA_DOSSIER.md` next to the recordings. Raw files are read-only: write cleaned or derived data elsewhere. The data contains patient identifiers, so do not share it.